### PR TITLE
Add check for `declaration` and `typeDefinition` support in vim lsp server before making `request`

### PR DESCRIPTION
**Pull request by kdheepak**
This PR is required because LanguageServer.jl crashes when requesting `textDocument/typeDefinition` and `textDocument/declaration`. LanguageServer.jl tells the client that it does not support this client and the client should not make this request. I've based this PR on the code for `textDocument/implementation`, which does the right thing.

### CODE REVIEW

**Review by teto — CHANGES_REQUESTED** [review]
this incondtionnaly disables typeDefinition. You need to really check for support in runtime/lua/vim/lsp/protocol.lua|863 col 19| function protocol.resolve_capabilities(server_capabilities)

**Comment by kdheepak** [comment]
Oh I didn't realize! I'll look into it now.

**Comment by kdheepak** [comment]
This is working as expected for me now. This is ready for review now.

File: runtime/lua/vim/lsp/protocol.lua [hunk 1]
```diff
@@ -923,6 +923,28 @@ function protocol.resolve_capabilities(server_capabilities)
     error("The server sent invalid codeActionProvider")
   end
 
+  if server_capabilities.declarationProvider == nil then
+    general_properties.declaration = false
+  elseif type(server_capabilities.declarationProvider) == 'boolean' then
+    general_properties.declaration = server_capabilities.declarationProvider
+  elseif type(server_capabilities.declarationProvider) == 'table' then
+    -- TODO: support more detailed declarationProvider options.
+    general_properties.declaration = false
```
Comment (tjdevries) [inline runtime/lua/vim/lsp/protocol.lua:932]: Hmm, when we do this, doesn't it disable these requests from happening? Should we default to true and see what happens instead?

Comment (kdheepak) [reply runtime/lua/vim/lsp/protocol.lua:932]: My understanding is that the block that deals with when `type(server_capabilities.declarationProvider) == table` hasn't been fully implemented yet. So maybe in a separate PR that can be redone properly.

For the other branches, I think neovim should not "default to `true` and see what happens". As an example, LanguageServer.jl specifies it does not support some features and throws an exception and crashes if a request is made that is not supported: https://github.com/julia-vscode/LanguageServer.jl/issues/736. I'm sure other language server have similar undefined behavior when this occurs.

Comment (kdheepak) [reply runtime/lua/vim/lsp/protocol.lua:932]: See the following for when `type(server_capabilities.implementationProvider) == table`, where it defaults to `false`.

https://github.com/kdheepak/neovim/blob/c37281f32fd49d1cada353d0b7ad95e42c1af0f2/runtime/lua/vim/lsp/protocol.lua#L954

Comment (tjdevries) [reply runtime/lua/vim/lsp/protocol.lua:932]: I think some of those are not exactly correct, but we could probably do as you suggested.

I am looking at the registration options and how they work, some of the things we do support (for example, linkSupport and others). This shouldn't be up to you to solve though probably :)

Comment (tjdevries) [reply runtime/lua/vim/lsp/protocol.lua:932]: When it returns a table, it means it does support the protocol, is my understanding. However, it gives additional information about certain registration options, incremental updates, etc.

Comment (kdheepak) [reply runtime/lua/vim/lsp/protocol.lua:932]: Yes, I agree, `table` does mean it supports the protocol and the options need to be parsed correctly.

Comment (tjdevries) [reply runtime/lua/vim/lsp/protocol.lua:932]: Are you interested in parsing the table or not for these requests? Just trying to gauge your interest :)

Comment (kdheepak) [reply runtime/lua/vim/lsp/protocol.lua:932]: I'd be interested in contributing! But as a separate PR? This currently crashes the Julia Language Server, and it'll be easier to convince colleagues to try to nightly nvim if that didn't happen :)

Comment (tjdevries) [reply runtime/lua/vim/lsp/protocol.lua:932]: :+1: Sounds good. Let me review again and I'll merge I think.

**Comment by tjdevries** [comment]
Thanks for the PR @kdheepak . Feel free to open another if you want to handle the more complicated parsing of the tables!

**Comment by kdheepak** [comment]
Thank you!

**Pull request merged**